### PR TITLE
add formatted timestamp variable

### LINKED ISSUE #9959: AMP Variable "TIMESTAMP" Format
I suppose a feature that it possible to format AMP variable 'TIMESTAMP'.

This is for displaying data as a character string for users when sending data with AMP Analytics.

For example, I think following usages.

```
AMP Variable: timestamp(yyyy-MM-dd'T'HH:mm:ss.SSSZ)
```

The format will be based on other language datetime format function, too.

### PR DESCRIPTION
**Pull request by tiendt**
Closes #9959

### CODE REVIEW

File: src/service/url-replacements-impl.js [hunk 1]
```diff
@@ -313,6 +313,20 @@ export class GlobalVariableSource extends VariableSource {
       return Date.now();
     });
 
+    //Returns the human readable timestamp in format of 2011-01-01 11:11:11.
```
Comment (cramforce) [inline src/service/url-replacements-impl.js]: The current code doesn't zero pad.

I'd highly prefer if we either

- don't do this
- or use `toISOString()` and be called `TIMESTAMP_ISO`

Formatting is highly arbitrary and people will want all kinds of format

Comment (tiendt) [inline src/service/url-replacements-impl.js]: @lannka what do you think?

Comment (lannka) [inline src/service/url-replacements-impl.js]: @cramforce that was my original recommendation, but requester wants time in local timezone. see conversation in #9959

File: test/functional/test-url-replacements.js [hunk 2]
```diff
@@ -374,6 +374,12 @@ describes.sandboxed('UrlReplacements', {}, () => {
     });
   });
 
+  it('should replace TIMESTAMP_FORMATTED', () => {
+    return expandAsync('?tsf=TIMESTAMP_FORMATTED').then(res => {
```
Comment (lannka) [inline test/functional/test-url-replacements.js]: please fake a time and test with real numbers.

File: src/service/url-replacements-impl.js (hunk 1 above)
Comment (cramforce) [inline src/service/url-replacements-impl.js]: Much better to use timezone offset then or to `toLocaleString`

Comment (lannka) [inline src/service/url-replacements-impl.js]: toLocaleString formats timestamp in different ways
for example, `en-US` gives `7/13/2017, 1:47:53 PM`
they will receive all different time formatting at server side :-).

how would the timezone offset look like?

File: test/functional/test-url-replacements.js (hunk 2 above)
Comment (tiendt) [inline test/functional/test-url-replacements.js]: @lannka I just updated the test, please take a look

File: test/functional/test-url-replacements.js [hunk 3]
```diff
@@ -1312,4 +1318,18 @@ describes.sandboxed('UrlReplacements', {}, () => {
           'NDrFP4BtPQJMyxE4jb9FDlp37OJL');
     });
   });
+
+  describe('formatted timestamp', () => {
+    it('should return correct formatted timestamp', () => {
+      const fakeTime = 1499979336612;
+      const offset_ = new Date(fakeTime).getTimezoneOffset() * 60000;
```
Comment (lannka) [inline test/functional/test-url-replacements.js]: add a comment why offset is needed

Comment (lannka) [inline test/functional/test-url-replacements.js]: `offset_` -> `offset`

File: test/functional/test-url-replacements.js [hunk 4]
```diff
@@ -1312,4 +1318,18 @@ describes.sandboxed('UrlReplacements', {}, () => {
...
+      sandbox.useFakeTimers(fakeTime + offset_);
+      return expandAsync(
+          '?tsf=TIMESTAMP_FORMATTED').then(res => {
+            expect(encodeURIComponent('2017-07-13 20:55:36'))
+                .to.equal('2017-07-13%2020%3A55%3A36');
+            expect(res).to.equal('?tsf=2017-07-13%2020%3A55%3A36');
```
Comment (lannka) [inline test/functional/test-url-replacements.js]: `return expect(expandAsync('?tsf=TIMESTAMP_FORMATTED')).to.eventually.equal('?tsf=' + encodeURIComponent('2017-07-13 20:55:36'))`

**Comment by tiendt** [comment]
@lannka also updated this one

File: src/service/url-replacements-impl.js [hunk 5]
```diff
@@ -552,9 +566,19 @@ export class GlobalVariableSource extends VariableSource {
...
   }
+
+    /**
+     * Format number to have 2 digits if number is smaller than 10
+     * @param num
+     * @returns {string} after formatted
+     * @private
+     */
+  formatNum_(num) {
+    return (num < 10 ? ('0' + num) : num);
```
Comment (jridgewell) [inline src/service/url-replacements-impl.js]: I think there are other uses who have something similar. A `leftPad` helper could be useful here.

Comment (jridgewell) [inline src/service/url-replacements-impl.js]: Nit: This doesn't always return a string.

File: src/string.js [hunk 6]
```diff
@@ -109,3 +109,26 @@ export function stringHash32(str) {
...
+/**
+ * Pads the current string with another string
+ * so that the returned string reaches the given length.
+ * The default value to pad with is " " if opt_padStr is not parsed
+ * @param {string} str
+ * @param {number} length
+ * @param {string=} opt_padStr
+ * @return {string}
+ */
+export function padStart(str, length, opt_padStr) {
```
Comment (lannka) [inline src/string.js]: see the `join` trick: https://stackoverflow.com/questions/10073699/pad-a-number-with-leading-zeros-in-javascript

File: test/functional/test-url-replacements.js [hunk 7]
```diff
@@ -374,6 +374,12 @@ describes.sandboxed('UrlReplacements', {}, () => {
     });
   });
 
+  it('should replace TIMESTAMP_FORMATTED', () => {
+    return expandAsync('?tsf=TIMESTAMP_FORMATTED').then(res => {
+      expect(res).to.match(/tsf=\d+/);
```
Comment (lannka) [inline test/functional/test-url-replacements.js]: seems this test is not needed , it's already covered by your other test.

File: src/service/url-replacements-impl.js (hunk 1 above)
Comment (cramforce) [inline src/service/url-replacements-impl.js]: `toISOString()` is what you want.

We should not do this `FORMATTED` thing. We need to at the very least adhere to some standard and put that in the name. Why is ISO bad?

Comment (cramforce) [inline src/service/url-replacements-impl.js]: TImezone is already in this file.

Comment (lannka) [inline src/service/url-replacements-impl.js]: Let's address this in the origin issue: #9959

**Comment by lannka** [comment]
as discussed in #9959, let's go with @cramforce's suggestion and make it a `TIMESTAMP_ISO `

**Comment by tiendt** [comment]
@lannka @cramforce please review the change to see if this is what expected

File: test/functional/test-url-replacements.js [hunk 8]
```diff
@@ -374,6 +374,20 @@ describes.sandboxed('UrlReplacements', {}, () => {
...
+    });
+  });
+
+  it('should return correct ISO timestamp', () => {
+    const fakeTime = 1499979336612;
+    sandbox.useFakeTimers(fakeTime);
+    return expect(expandAsync('?tsf=TIMESTAMP_ISO'))
+        .to.eventually.equal(
+        '?tsf=' + encodeURIComponent('2017-07-13T20:55:36.612Z'));
```
Comment (cramforce) [inline test/functional/test-url-replacements.js]: Typically better to avoid logic in tests and just put the literal value.

Comment (lannka) [inline test/functional/test-url-replacements.js]: +1

**Comment by tiendt** [comment]
@lannka can you take a look at the new change and possibly merge this? Thanks!

**Pull request merged**